eclipse-theia/theia
# fix user storage regressions

#### What it does


- fix #8302: ensure that user storage URI has a top-level directory. Otherwise `FileService` fails to create the user directory.
- remove hardcoded constants from preferences UI. It is a secret of the preference service to know which config URI should be use for each scope in the context of the given resource.
- fix #8301: use monaco model to handle keymaps. It allows to check whether model is invalid before opening it, in addition keymaps are now updated with minimal edits and preserving user formatting and comments.

#### How to test


- Start theia app and remove .theia in user home. Try to open keymaps and user settings.
- Try to modify keymaps and check that changes are reflected.
- Try to open settings file for different scopes, i.e. user, workspace, folder as a workspace and workspace folder.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
